circuit/src/integer/mod.rs

```diff
@@ -250,100 +91,611 @@ where
 		for i in 0..NUM_LIMBS {
 			assigned_result[i] = Some(region.assign_advice(
 				|| format!("result_{}", i),
-				config.x_limbs[i],
+				common.advice[i + NUM_LIMBS],
 				row + 1,
 				|| Value::known(reduction_witness.result.limbs[i]),
 			)?)
 		}
-		let assigned_result = assigned_result.map(|x| x.unwrap());
+		//let assigned_result = assigned_result.map(|x| x.unwrap();
+		let assigned_result = AssignedInteger::new(
+			&reduction_witness.result,
+			&assigned_result.map(|x| x.unwrap()),
+		);
 		Ok(assigned_result)
 	}
+}
+
+/// Chip structure for the IntegerReduce.
+pub struct IntegerReduceChip<
+	W: FieldExt,
+	N: FieldExt,
+	const NUM_LIMBS: usize,
+	const NUM_BITS: usize,
+	P,
+> where
+	P: RnsParams<W, N, NUM_LIMBS, NUM_BITS>,
+{
+	// Assigned integer
+	assigned_integer: AssignedInteger<W, N, NUM_LIMBS, NUM_BITS, P>,
+	/// Constructs phantom datas for the variables.
+	_native: PhantomData<N>,
+	_wrong: PhantomData<W>,
+	_rns: PhantomData<P>,
+}
+
+impl<W: FieldExt, N: FieldExt, const NUM_LIMBS: usize, const NUM_BITS: usize, P>
+	IntegerReduceChip<W, N, NUM_LIMBS, NUM_BITS, P>
+where
+	P: RnsParams<W, N, NUM_LIMBS, NUM_BITS>,
+{
+	/// Creates a new reduce chip
+	pub fn new(assigned_integer: &AssignedInteger<W, N, NUM_LIMBS, NUM_BITS, P>) -> Self {
+		Self {
+			assigned_integer: assigned_integer.clone(),
+			_native: PhantomData,
+			_wrong: PhantomData,
+			_rns: PhantomData,
+		}
+	}
+}
 
-	/// Assign cells for reduce operation.
-	pub fn reduce(
-		x_limbs: [AssignedCell<N, N>; NUM_LIMBS],
-		rw: ReductionWitness<W, N, NUM_LIMBS, NUM_BITS, P>, config: IntegerConfig<NUM_LIMBS>,
-		mut layouter: impl Layouter<N>,
-	) -> Result<[AssignedCell<N, N>; NUM_LIMBS], Error> {
+impl<W: FieldExt, N: FieldExt, const NUM_LIMBS: usize, const NUM_BITS: usize, P> Chip<N>
+	for IntegerReduceChip<W, N, NUM_LIMBS, NUM_BITS, P>
+where
+	P: RnsParams<W, N, NUM_LIMBS, NUM_BITS>,
+{
+	type Output = AssignedInteger<W, N, NUM_LIMBS, NUM_BITS, P>;
+
+	/// Make the circuit config.
+	fn configure(common: &CommonConfig, meta: &mut ConstraintSystem<N>) -> Selector {
+		let selector = meta.selector();
+		let p_in_n = P::wrong_modulus_in_native_modulus();
+
+		meta.create_gate("reduce", |v_cells| {
+			let s = v_cells.query_selector(selector);
+			let mut t_exp = [(); NUM_LIMBS].map(|_| None);
+			let reduce_q_exp = v_cells.query_advice(common.advice[NUM_LIMBS], Rotation::prev());
+			let mut limbs_exp = [(); NUM_LIMBS].map(|_| None);
+			let mut residues_exp = Vec::new();
+			let mut result_exp = [(); NUM_LIMBS].map(|_| None);
+			for i in 0..NUM_LIMBS {
+				t_exp[i] = Some(v_cells.query_advice(common.advice[i], Rotation::prev()));
+				limbs_exp[i] = Some(v_cells.query_advice(common.advice[i], Rotation::cur()));
+				if i < NUM_LIMBS / 2 {
+					residues_exp.push(v_cells.query_advice(common.advice[i], Rotation::next()));
+				}
```

Comment:
Can we move this out of this loop and put it inside its own loop, just to avoid nesting and make it more clear.
Apply in all similar cases